File: Arrays/2016. Maximum Difference Between Increasing Elements/2016. Maximum Difference Between Increasing Elements.py
class Solution:
    def maximumDifference(self, nums: List[int]) -> int:
        max_diff = -1
        min_element = nums[0]

        for i in range(1, len(nums)):
            if nums[i] == min_element:
                continue

            if nums[i] < min_element:
                min_element = nums[i]
            else:
                max_diff = max(max_diff, nums[i] - min_element)
        
        return max_diff

# Time: O(n)
# Space: O(1)

File: Arrays/2016. Maximum Difference Between Increasing Elements/2016. Maximum Difference Between Increasing Elements.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumDifference = function (nums) {
  let max_diff = -1;
  let min_element = nums[0];

  for (let i = 1; i < nums.length; i++) {
    if (nums[i] === min_element) continue;

    if (nums[i] < min_element) min_element = nums[i];
    else max_diff = Math.max(max_diff, nums[i] - min_element);
  }

  return max_diff;
};

// Time: O(n)
// Space: O(1)
